xlators/storage/posix/src/posix-helpers.c

```diff
@@ -2345,78 +2342,105 @@ posix_disk_space_check(xlator_t *this)
-    LOCK(&priv->lock);
-    {
-        priv->disk_space_check_active = _gf_false;
+                THIS = this = pthis->xl;
+                priv = this->private;
+
+                posix_disk_space_check(priv);
+
+                pthread_mutex_lock(&ctx->xl_lock);
+                {
+                    pthis->is_use = _gf_false;
+                    /* Send a signal to posix_notify function */
+                    if (pthis->detach_notify)
+                        pthread_cond_signal(&pthis->cond);
+                }
+            }
+
+            timespec_now_realtime(&sleep_till);
+            sleep_till.tv_sec += 5;
+            (void)pthread_cond_timedwait(&ctx->xl_cond, &ctx->xl_lock,
+                                         &sleep_till);
+        }
+        pthread_mutex_unlock(&ctx->xl_lock);
     }
-    UNLOCK(&priv->lock);
 
     return NULL;
 }
 
 int
-posix_spawn_disk_space_check_thread(xlator_t *xl)
+posix_spawn_disk_space_check_thread(xlator_t *this)
 {
-    struct posix_private *priv = NULL;
-    int ret = -1;
+    int ret = 0;
+    glusterfs_ctx_t *ctx = NULL;
+    struct posix_diskxl *pxl = NULL;
+    struct posix_private *priv = this->private;
 
-    priv = xl->private;
+    ctx = this->ctx;
 
-    LOCK(&priv->lock);
+    pthread_mutex_lock(&ctx->xl_lock);
     {
-        /* cancel the running thread  */
-        if (priv->disk_space_check_active == _gf_true) {
-            pthread_cancel(priv->disk_space_check);
-            priv->disk_space_check_active = _gf_false;
-        }
+        if (ctx->diskxl_count++ == 0) {
+            ret = gf_thread_create(&ctx->disk_space_check, NULL,
+                                   posix_ctx_disk_thread_proc, ctx,
+                                   "posixctxres");
 
-        ret = gf_thread_create(&priv->disk_space_check, NULL,
-                               posix_disk_space_check_thread_proc, xl,
-                               "posixrsv");
-        if (ret) {
-            priv->disk_space_check_active = _gf_false;
-            gf_msg(xl->name, GF_LOG_ERROR, errno, P_MSG_DISK_SPACE_CHECK_FAILED,
-                   "unable to setup disk space check thread");
-            goto unlock;
+            if (ret) {
+                gf_msg(this->name, GF_LOG_ERROR, errno, P_MSG_THREAD_FAILED,
+                       "spawning disk space check thread failed");
+                ctx->diskxl_count--;
+                pthread_mutex_unlock(&ctx->xl_lock);
+                goto out;
+            }
         }
-
-        priv->disk_space_check_active = _gf_true;
+        pxl = calloc(1, sizeof(*pxl));
+        if (!pxl) {
+            ret = ENOMEM;
+            gf_log(this->name, GF_LOG_ERROR,
+                   "Calloc is failed to allocate "
+                   "memory for diskxl object");
+            pthread_mutex_unlock(&ctx->xl_lock);
+            goto out;
+        }
+        pthread_cond_init(&pxl->cond, NULL);
+        pxl->xl = this;
+        priv->pxl = (void *)pxl;
```

Comment:
All the initialization should be done before taking the mutex. Inside the mutex we only need to add the entry to the list.